explorer/src/ingest.js

```diff
@@ -0,0 +1,224 @@
+#!/usr/bin/env node
+
+const args = require("args-parser")(process.argv);
+const { ApiPromise, WsProvider } = require("@polkadot/api");
+const types = require(`${process.cwd()}/blockchain/types.json`);
+const { Client: PostgresClient } = require('pg');
+
+async function main() {
+  const { chain, postgres: postgresUrl } = args;
+  if (!chain || !postgresUrl) {
+    console.log("Please provide --chain=wss://yourchain and --postgres=postgres://some-postgres");
+    process.exit(1);
+  }
+
+  const pgClient = new PostgresClient({
+    connectionString: postgresUrl,
+    ssl: {
+      rejectUnauthorized: false
+    }
+  });
+  await pgClient.connect();
+
+  const storage = await PostgresStorage.create(pgClient);
+
+  const provider = new WsProvider(chain);
+  const api = await ApiPromise.create({
+    provider,
+    types,
+  });
+
+  updateLargestBlock(api, storage);
+
+  while (true) {
+    const didWork = await catchUpToLatest(api, storage);
+    if (!didWork) {
+      await sleep(6000);
+    }
+  }
+}
+
+main()
+  .then(() => process.exit(0))
+  .catch(err => {
+    console.error(err);
+    process.exit(1);
+  });
+
+class PostgresStorage {
+  constructor(postgres) {
+    this.pg = postgres;
+  }
+
+  static async create(postgres) {
+    await postgres.query(`
+      CREATE TABLE IF NOT EXISTS
+      key_values (
+        key VARCHAR PRIMARY KEY,
+        value VARCHAR NOT NULL
+      )
+    `);
+    return new PostgresStorage(postgres);
+  }
+
+  async fullyIngested() {
+    const val = await this.getKey('ingestion/fully_ingested');
+    return val && Number(val);
+  }
+
+  async getKey(key) {
+    const val = await this.pg.query(`SELECT value FROM key_values WHERE key = '${key}' LIMIT 1`);
+    return val.rows[0]?.value;
+  }
+
+  async setKey(key, value) {
+    await this.pg.query(`
+      INSERT INTO key_values (key, value) VALUES ('${key}', '${value}')
+      ON CONFLICT (key) DO UPDATE SET value='${value}'
+    `);
+  }
+
+  async setFullyIngested(number) {
+    await this.setKey('ingestion/fully_ingested', number.toString());
+  }
+
+  async setLargestBlock(number) {
+    await this.setKey('ingestion/largest', number.toString());
+  }
+
+  async saveBlock(number, blockData) {
+    const json = JSON.stringify(blockData);
+    await this.setKey(`block/${number}`, json);
+  }
+
+  async saveExtrinsic(blockNumber, index, extrinsicData) {
+    const json = JSON.stringify(extrinsicData);
+    await this.setKey(`block/${blockNumber}/extrinsic/${index}`, json);
+  }
+}
+
+async function updateLargestBlock(api, storage) {
+  while (true) {
+    const latestBlockNumber = (await api.rpc.chain.getBlock()).block.header.number.toNumber();
+    await storage.setLargestBlock(latestBlockNumber);
+    console.log(`Largest block ${latestBlockNumber}`)
+    await sleep(6000);
```

Comment:
Maybe make the sleep duration configurable.

Reply:
I made it not sleep at all, and have it do work as soon as there's a new block. Much better solution than sleeping at all.